controller/volume_attachment_controller.go

```diff
@@ -283,129 +292,192 @@ func (vac *VolumeAttachmentController) handleVolumeAttachment(va *longhorn.Volum
 		return
 	}
 
-	attachment := selectAttachmentToAttach(va)
-	if attachment == nil {
+	attachmentSpec := selectAttachmentSpecToAttach(va, vol)
+	if attachmentSpec == nil {
 		return
 	}
 
-	vol.Spec.NodeID = attachment.NodeID
-	setAttachmentParameter(attachment, vol)
+	vol.Spec.NodeID = attachmentSpec.NodeID
+	setAttachmentParameter(attachmentSpec.Parameters, vol)
 	return
 }
 
-func selectAttachmentToAttach(va *longhorn.VolumeAttachment) *longhorn.Attachment {
-	if len(va.Spec.Attachments) == 0 {
+func selectAttachmentSpecToAttach(va *longhorn.VolumeAttachment, vol *longhorn.Volume) *longhorn.AttachmentSpec {
+	// TODO: consider handle ignore the csi-attacher ticket for RWX volume
+	if len(va.Spec.AttachmentSpecs) == 0 {
 		return nil
 	}
 
-	highPriorityAttachments := []*longhorn.Attachment{}
+	highPriorityAttachmentSpecs := []*longhorn.AttachmentSpec{}
 	maxAttacherPriorityLevel := 0
-	for _, attachment := range va.Spec.Attachments {
-		priorityLevel := longhorn.GetAttacherPriorityLevel(attachment.Type)
+	for _, attachmentSpec := range va.Spec.AttachmentSpecs {
+		// TODO: refactor this! Consider handle ignore the csi-attacher ticket for RWX volume
+		if vol.Spec.AccessMode == longhorn.AccessModeReadWriteMany &&
+			!vol.Spec.Migratable &&
+			attachmentSpec.Type == longhorn.AttacherTypeCSIAttacher {
+			continue
+		}
+		priorityLevel := longhorn.GetAttacherPriorityLevel(attachmentSpec.Type)
 		if priorityLevel > maxAttacherPriorityLevel {
 			maxAttacherPriorityLevel = priorityLevel
 		}
 	}
 
-	for _, attachment := range va.Spec.Attachments {
-		priorityLevel := longhorn.GetAttacherPriorityLevel(attachment.Type)
+	for _, attachmentSpec := range va.Spec.AttachmentSpecs {
+		// TODO: refactor this! Consider handle ignore the csi-attacher ticket for RWX volume
+		if vol.Spec.AccessMode == longhorn.AccessModeReadWriteMany &&
+			!vol.Spec.Migratable &&
+			attachmentSpec.Type == longhorn.AttacherTypeCSIAttacher {
+			continue
+		}
```

Comment:
NIT: Can we have a negative priority for csi-attacher tickets of RWX volumes? Then the controller does not need to do this.

Reply:
We would still have to check and exclude the negative ticket. I created the function `isCSIAttacherTicketOfRegularRWXVolume` to simplify the current checking logic. I think it is more verbal about why we exclude these rwx csi tickets. What do you think?